compiler/rustc_hir/src/hir.rs

```diff
@@ -644,13 +644,41 @@ impl<'hir> Generics<'hir> {
         })
     }
 
-    pub fn bounds_span_for_suggestions(&self, param_def_id: LocalDefId) -> Option<Span> {
+    /// Returns bounds span for suggestions.
+    /// If the span including lifetime bound needs parentheses, it returns a span to the open parenthese at the second item.
+    /// e.g. `dyn Future<Output = ()> + 'static` needs parentheses `(dyn Future<Output = ()>) + 'static`
+    pub fn bounds_span_for_suggestions(
+        &self,
+        param_def_id: LocalDefId,
+    ) -> Option<(Span, Option<Span>)> {
         self.bounds_for_param(param_def_id).flat_map(|bp| bp.bounds.iter().rev()).find_map(
             |bound| {
-                // We include bounds that come from a `#[derive(_)]` but point at the user's code,
-                // as we use this method to get a span appropriate for suggestions.
-                let bs = bound.span();
-                bs.can_be_used_for_suggestions().then(|| bs.shrink_to_hi())
+                let span_for_parentheses = if let Some(trait_ref) = bound.trait_ref()
+                    && let [.., segment] = trait_ref.path.segments
+                    && segment.args().parenthesized == GenericArgsParentheses::ParenSugar
+                    && let [binding] = segment.args().bindings
+                    && let TypeBindingKind::Equality { term: Term::Ty(ret_ty) } = binding.kind
```

Comment:
Ah, I think you need to [`peel_refs`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.Ty.html#method.peel_refs) here, so that we detect things like `Fn() -> &'static dyn Future<Output = ()>` (see my other review comments for examples).

```suggestion
                    && let TypeBindingKind::Equality { term: Term::Ty(ret_ty) } = binding.kind
                    && let ret_ty = ret_ty.peel_refs()
```